pkg/services/baremetal/baremetal/baremetal.go

```diff
@@ -363,77 +366,121 @@ func (s *Service) getAssociatedHost(ctx context.Context) (*infrav1.HetznerBareMe
 	return &host, helper, nil
 }
 
-func (s *Service) chooseHost(ctx context.Context) (*infrav1.HetznerBareMetalHost, *patch.Helper, error) {
+// chooseHost tries to find a free hbmh.
+// If no hbmh was found, then hbmh and err are nil, and the string
+// "reason" contains human readable details.
+func (s *Service) chooseHost(ctx context.Context) (
+	hbmh *infrav1.HetznerBareMetalHost, patchHelper *patch.Helper, reason string, err error,
+) {
 	// get list of hosts scoped to namespace of machine
 	hosts := infrav1.HetznerBareMetalHostList{}
 	opts := &client.ListOptions{
 		Namespace: s.scope.BareMetalMachine.Namespace,
 	}
 
 	if err := s.scope.Client.List(ctx, &hosts, opts); err != nil {
-		return nil, nil, fmt.Errorf("failed to list hosts: %w", err)
+		return nil, nil, "", fmt.Errorf("failed to list hosts: %w", err)
 	}
 
 	labelSelector := s.getLabelSelector()
 
 	availableHosts := make([]*infrav1.HetznerBareMetalHost, 0, len(hosts.Items))
 
+	mapOfSkipReasons := make(map[string]int)
+
+	usableDevices := 0
 	for i, host := range hosts.Items {
 		if host.Spec.ConsumerRef != nil && consumerRefMatches(host.Spec.ConsumerRef, s.scope.BareMetalMachine) {
 			helper, err := patch.NewHelper(&hosts.Items[i], s.scope.Client)
 			if err != nil {
-				return nil, nil, fmt.Errorf("failed to create patch helper: %w", err)
-			}
-			return &hosts.Items[i], helper, nil
-		}
-		if s.scope.BareMetalMachine.Spec.InstallImage.Swraid == 1 {
-			// Machine should have RAID. Skip machines which have less than two WWNs
-			if len(host.Spec.RootDeviceHints.Raid.WWN) < 2 {
-				continue
+				return nil, nil, "", fmt.Errorf("failed to create patch helper: %w", err)
 			}
+			return &hosts.Items[i], helper, "", nil
 		}
 		if host.Spec.ConsumerRef != nil {
 			continue
 		}
+		if host.Spec.Status.ProvisioningState != infrav1.StateNone {
+			continue
+		}
+
+		// from now on each "continue" should add an entry
+		// to mapOfSkipReasons.
+		usableDevices++
+
+		if host.Spec.RootDeviceHints != nil &&
```

Comment:
this code block has to come last. If the host is in maintenance mode, then getting this reason is misleading

Reply:
makes sense. Done in 40f22534